### DISCUSSION TITLE
How to pixelate images, without resize (Nearest Neighbour)

### DISCUSSION THREAD

**Discussion by Intenzi** [python-pillow/Pillow#7660] [Q&A] [answered]
Hello, this is a short question and I believe it does not require any code example. Please do read, thankyou!
I am looking to utilise the python equivalent of css's `image-rendering: pixelated`

My current workaround is to utilise `Image.resize()` with Image.Resampling.NEAREST as the filter
Which works properly.

### The issue:
I currently resize to a fixed size for all images.
However, this does not work for images that are the same size as the size I specify for resizing.. (they don't get pixelated)

I have tried resizing the images to a larger size using `Image.resize` and then resizing back with nearest neighbour filter, but it does not render the pixelated image.
Thus, how can I perform resampling with nearest neighbour on images that I can't resize, as their dimensions are the same as I require.
Should I just go for a gimmick currently by utilising +1 width and +1 height so that `Image.resize` occurs and performs the resampling?

**Comment by radarhere** [comment]
Have you tried resizing them to a **smaller** size and then resizing them back?

**Reply by Intenzi** [reply]
Hi there, yes I have tried resizing them to a smaller size @radarhere 
It does not lead to anything substantial however and is still blurry.

Take the following image as an example -> https://play.pokemonshowdown.com/sprites/gen5-back/venusaur-mega.png
![image](https://github.com/python-pillow/Pillow/assets/62426443/31dc40e7-8e78-41e7-85af-a090c4ab8eea)
When you go to its link, it is blurry. However upon using the css property of `image-rendering: pixelated` it gets visibly more defined.
![image](https://github.com/python-pillow/Pillow/assets/62426443/e7f8f1fb-5322-4981-893a-47ba6f815e8e)

I used the two crown pointy shapes at the back(arrowed above) in the pixelated version as a reference to quality.
(After zooming in on each image, as without zooming in, everything looks very similar)
Which I then compared with various other renderings of the image via PIL.
The image dimensions are 96x96.

Initial code:-
```py
from PIL import Image
i = Image.open('venusaur-mega.png').convert('RGBA')
```

1. Upon running `i.resize((96,96), Image.Resampling.NEAREST)).show()`
There is no visible change that can be observed.
![tmpf1xin_5v](https://github.com/python-pillow/Pillow/assets/62426443/0ae84bbe-cd9c-40de-b7ae-caa560142cc8)

2. Upon running `i.resize((144,144), Image.Resampling.NEAREST)).show()`
There is a visible change that can be observed. 
Which is **very similar to the *96x96 image*** on the website link at the very top, after using the css property of `image-rendering: pixelated`

![tempbigsize](https://github.com/python-pillow/Pillow/assets/62426443/87cbb6e5-842c-4ec5-aed8-6a048ef2728f)

A quick zoomed in comparison of (the first run vs the second run)
![image](https://github.com/python-pillow/Pillow/assets/62426443/83193db6-16a1-4e81-ac55-6817bda799bf)

Any further attempts such as downsizing and then resizing back to 96x96
or upsizing and resizing back to 96x96 are similar to the base image.

I don't know how nearest neighbour resampling works, or whether it is actually occurring properly upon downsizing and resizing and just isn't shown as clearly because of my comparison being after resizing to a bigger image (144x144).
However, if I compare them with the 96x96 image I can see on the website link after using the pixelated rendering (2nd attachment), it is blurry in comparison.

Thankyou for reading this!

**Reply by nulano** [reply]
That looks like the image is blurry not because of Pillow but your image viewer.

What do you think of the following enlarging your image to 16x the size?

```python
from PIL import Image
im = Image.open(r"C:\Users\Nulano\Downloads\venusaur-mega.png").convert("RGBA")
im = im.resize((im.size[0] * 16, im.size[1] * 16), Image.Resampling.BOX)
im.show()
```

This does not look blurry for me:

![image](https://github.com/python-pillow/Pillow/assets/3819630/3c62fcdf-629f-44b3-8168-929d7b695996)

To explain [`Image.Resampling.BOX`](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#PIL.Image.Resampling.BOX), it is the same as `Image.Resampling.NEAREST` when upscaling, but you might find it better when downscaling.

**Reply by Intenzi** [reply]
Hi @nulano , we're both using the same image viewer on windows.

You are absolutely correct that the image quality becomes not blurry after resampling is done with an increase in size!
And I am infact using it as mentioned in my original comment and it works very similar to the css `image-rendering: pixelated` property.

It seems you got my query slightly wrong.
I am unable to figure how to pixelate the image when I do not want an increase in size. There is a very major difference when using and not using the image rendering property, which you can try out in the image link I sent (in my above reply) along with a snapshot of what to modify using inspect element.

I have already mentioned that resizing it back to original size after increasing size does not make it appear pixelated/clear.

Here is a screenshot of comparison after uploading them to discord.
In order:-
1. Normal downloaded image 96x96
2. im.resize(im.size[0] * 16, im.size[1] * 16, Image.Resampling.BOX).resize((96, 96), Image.Resampling.BOX).save("other.png")
3. im.resize((144, 144), Image.Resampling.NEAREST)
![image](https://github.com/python-pillow/Pillow/assets/62426443/fe22d575-1e7e-435e-90a5-f2c8ccec8c3b)

I hope this clears up that isn't the image viewer's issue and that the pixels do look extremely different with a rendering after upsize.
And that they look nearly the same as original/not defined at all upon upsizing and then resizing back to original size.
Downsizing and resizing back to original size degrades the pixels in quality severely and doesn't seem to be helpful.

**Reply by nulano** [reply]
I think you may be confused on what resampling with nearest does.

If you upscale an image by an integer factor (say 32x32 to 96x96, i.e. 3x larger), it simply repeats each pixel value 3 times.
So instead of having pixels (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, ...) as in the original image, you have (0, 0, 0, 1, 1, 1, 2, 2, 2, ...) in the upscaled image.

So it makes sense that if you keep the same size, you get the same image.

There have been two suggestions:

* downscale the image, then upscale it: When you downscale the image, you get a new image with some values. Upscaling this image will again produce repeated pixels which gives you a pixelated look. However, as you know, the downscaling step will decrease quality.
* upscale the image, then downscale it again: Depending on the scaling method you use, you either get back the original image or a blurry version of it

The reason upscaling an image makes it look more "pixelated" is that there are repeated pixels, so when an image viewer (which will typically use something like bilinear resampling) upscales it again, it creates sharper edges. However, if you want to keep the original size, there is no way to tell an image viewer to create these sharp edges.

**Reply by Intenzi** [reply]
heya @nulano, thanks for the reply
sorry for the late response, happy new year!

The two suggested options for obtaining a clearer pixel image without resizing:-
- `downscale the image, then upscale it:` It does not only decrease quality, it gets worse than existing downscaled image.
- `upscale the image, then downscale it again:` Yes, it only gives back the original image or a blurry version of it.
which implies that its not possible..

HOWEVER, it is also a fact that it is possible without resizing when using the css property of `image-rendering: pixelated`
Even though, the mdn docs for it say that it must be scaled to work, it works regardless?

**I tried looking up at *how* this works..**
Source: https://developer.chrome.com/blog/pixelated
> The CSS property image-rendering and the value pixelated are interesting because they turn off the browser's standard smooth scaling (normally bi-linear interpolation) and replaces it with another scaling algorithm (nearest neighbor in most cases) when resizing the images.

### Running a bit deeper into the rabbit hole, I believe I have found how it actually works.
As you had mentioned, nearest neighbour doesn't work without scaling in size.
However, the pixelated css seems to still perform a better visual render when used, which is really a big mystery to me.

This appears to be happening as `pixelated` not only performs nearest neighbour, it also does smooth rendering?
Source: https://drafts.csswg.org/css-images/#valdef-image-rendering-pixelated
> pixelated:
> The image is scaled in a way that preserves the pixelation of the original as much as possible, but allows minor smoothing as necessary to avoid distorting the image when the target size is not a clean multiple of the original.
> For each axis independently, first determine the integer multiple of its natural size that puts it closest to the target size and is greater than zero. Scale it to this integer-multiple-size using [nearest neighbor](https://drafts.csswg.org/css-images/#nearest-neighbor), then scale it the rest of the way to the target size as for [smooth](https://drafts.csswg.org/css-images/#valdef-image-rendering-smooth).

It was an interesting read to know about pixelated and the further examples they give after scrolling of comparisons of crisp-edges and pixelated.
> For most purposes, pixelated will produce a more suitable rendering than nearest neighbor.

### I am now wondering, how to re-enact `pixelated` with Pillow?
Definitely just using resize with Nearest Neighbour isn't sufficient, and it also needs to use smooth, which should be a filter as it doesn't depend on scaling an image? However in the linked source, it mentions smooth requires scaling as well 🤔

It has also mentioned that for scaling with integer multiples of the natural size, it performs same as crisp-edges.
Just a wild guess, that even crisp-edges can work over here as we can take the integer multiple to be 1 in this case..

Any help would be appreciated in figuring this out!
PS: Should I create a new discussion thread as the question seems to have changed with all of this context?

**Reply by nulano** [reply]
When you use the css property `image-rendering: pixelated`, you are telling every browser to use nearest neighbour scaling for rendering the image. So if the image has a different size than what is required to fill the HTML element with the image, it is scaled with nearest neigbour scaling, instead of the default.

AFAIK Windows Photos has no way to change its default scaling, so it will always make pixel art look blurry when zoomed in.

Similarly, if you don't control Discord's CSS, it will use the default scaling method to upscale images (which, based on your screenshot, it does by default for small images).

Since you can't change Windows Photos nor Discord's CSS with Pillow, it seems what you are asking for is impossible (i.e. changing an image's default upscaling method without changing its size).

---

The smoothing you mentioned is a red herring - it is talking about scaling by a non-integer factor, e.g. 32x32 to 40x40.
I don't know how it is implemented so I can't comment on how helpful it is, but it is not something that is implemented in Pillow AFAIK.

**Reply by Intenzi** [reply]
Hi @nulano I am confused by your response/?
I will take your word that changing an image's default upscaling method without changing its size is impossible
**Thus, it makes me unable to comprehend how an image on a website can do that?**

I am not going to be utilizing the image solely on its own. It gets pasted to a background and then gets uploaded anywhere such as discord/web/viewed on windows photos.
And over there, pasting the image to the background after changing its upscaling method, makes a stark difference.

![image](https://github.com/python-pillow/Pillow/assets/62426443/1eb97cc6-20b1-466e-8f9a-c6beec3df0d0)
As you can see, both are images uploaded to discord, the images themselves in css of discord will ofcourse not be utilizing nearest neighbour interpolation and whatever is the standard interpolation of the web or has been set by discord's css will get used.
However, the stark difference exists. **That is controlled by Pillow beforehand. It does not depend on what software I am using to view.**

**Reply by nulano** [reply]
By changing the CSS, you change the default upscaling method (for the browser, when displaying that particular image). It is not a property of the image.

> That is controlled by Pillow beforehand. It does not depend on what software I am using to view.

The only way to change this (i.e. achieve the difference between those two images) is by resizing the image. If you are worried about the image not fitting the background, remember that you can also resize the background.

Edit: For a fair comparison of sharpness, you could look at the image at 100% zoom (number in the bottom right of Windows Photos; not sure if you can do it in Discord). I imagine that those two images will have a very different size, but both will look perfectly sharp.

**Reply by Intenzi** [reply]
Thankyou for clearing my doubts!
I appreciate it a lot ^^

**Reply by Intenzi** [reply]
Pillow does not support changing the default upscaling method I believe?
Could I use something like photoshop to change the default upscaling method of my images.

> The only way to change this (i.e. achieve the difference between those two images) is by resizing the image. If you are worried about the image not fitting the background, remember that you can also resize the background.

Do you perhaps have any insights on how the css is managing to do this then, without even resizing the images at 100% zoom level? Maybe it can be implemented into Pillow :>

**Reply by nulano** [reply, marked as answer]
When a browser is displaying an image, it first calculates the size that is required to fill the HTML image tag. Usually this is the same as the image size, but it can be different (e.g. in Discord).

If the image size is different that the size of the HTML image tag, the browser has to resize the image to display it.
The CSS tag is simply telling the browser which scaling method to use instead of the default.

So you cannot change the default scaling method with *any* image editor, you can only do it by changing the CSS.

---

The workaround you found with upscaling the image before adding it to a page works by changing how much resizing the browser has to perform to fill the HTML tag (less resizing with the default method gives sharper pixel art).
